### PR TITLE
fix: fix config loading compatibility on Windows

### PR DESCRIPTION
**Pull request by chawyehsu**
- Fix https://github.com/soywod/himalaya/issues/72#issuecomment-822101907
- Use `USERPROFILE` as alternative HOME on Windows

### CODE REVIEW

File: src/config/model.rs [hunk 1]
```diff
@@ -47,7 +47,7 @@ impl Account {
 
     pub fn imap_passwd(&self) -> Result<String> {
         let passwd = run_cmd(&self.imap_passwd_cmd).chain_err(|| "Cannot run IMAP passwd cmd")?;
-        let passwd = passwd.trim_end_matches("\n").to_owned();
+        let passwd = passwd.trim_end().to_owned();
```
Comment (soywod) [inline src/config/model.rs:50]: With `trim_end` you also remove spaces, which are *in theory* valid chars for a password. I admit that *in practice* nobody uses this char in his password, but still. What do you think about:

```rust
let passwd = passwd
    .trim_end_matches(|c| c == '\r' || c == '\n')
    .to_owned();
```

File: src/config/model.rs [hunk 2]
```diff
@@ -68,7 +68,7 @@ impl Account {
 
     pub fn smtp_creds(&self) -> Result<SmtpCredentials> {
         let passwd = run_cmd(&self.smtp_passwd_cmd).chain_err(|| "Cannot run SMTP passwd cmd")?;
-        let passwd = passwd.trim_end_matches("\n").to_owned();
+        let passwd = passwd.trim_end().to_owned();
```
Comment (soywod) [inline src/config/model.rs:71]: With `trim_end` you also remove spaces, which are *in theory* valid chars for a password. I admit that *in practice* nobody uses this char in his password, but still. What do you think about:

```rust
let passwd = passwd
    .trim_end_matches(|c| c == '\r' || c == '\n')
    .to_owned();
```

File: src/config/model.rs [hunk 3]
```diff
@@ -115,7 +115,9 @@ impl Config {
     }
 
     fn path_from_xdg_alt() -> Result<PathBuf> {
-        let path = env::var("HOME").chain_err(|| "Cannot find `HOME` env var")?;
+        let path = env::var("HOME")
+            .or(env::var("USERPROFILE"))
+            .chain_err(|| "Cannot find `HOME` env var")?;
```
Comment (soywod) [inline src/config/model.rs:120]: `USERPROFILE` is only used by Windows, right? So I will go with sth more concise, like:

```rust
let home_var = if cfg!(target_family = "windows") {
    "USERPROFILE"
} else {
    "HOME"
};
let home_path: PathBuf = env::var(home_var)
    .chain_err(|| format!("Cannot find `{}` env var", home_var))?
    .into();
```

See https://doc.rust-lang.org/reference/conditional-compilation.html#target_family.

File: src/config/model.rs [hunk 4]
```diff
@@ -125,7 +127,9 @@ impl Config {
     }
 
     fn path_from_home() -> Result<PathBuf> {
-        let path = env::var("HOME").chain_err(|| "Cannot find `HOME` env var")?;
+        let path = env::var("HOME")
+            .or(env::var("USERPROFILE"))
+            .chain_err(|| "Cannot find `HOME` env var")?;
```
Comment (soywod) [inline src/config/model.rs:132]: `USERPROFILE` is only used by Windows, right? So I will go with sth more concise, like:

```rust
let home_var = if cfg!(target_family = "windows") {
    "USERPROFILE"
} else {
    "HOME"
};
let home_path: PathBuf = env::var(home_var)
    .chain_err(|| format!("Cannot find `{}` env var", home_var))?
    .into();
```

See https://doc.rust-lang.org/reference/conditional-compilation.html#target_family.

**Review by soywod — CHANGES_REQUESTED** [review]
Thanks for your contribution! I had some tiny changes, let me know what you think about them.

File: src/config/model.rs (hunk 1 above)
Comment (chawyehsu) [reply src/config/model.rs:50]: Make sense, I will update the patch.

**Review by soywod — APPROVED** [review]

**Comment by soywod** [comment]
Thanks :tada:

**Pull request merged**